src/mono/mono/metadata/marshal.c

```diff
@@ -3702,14 +3702,23 @@ mono_marshal_get_native_wrapper (MonoMethod *method, gboolean check_exceptions,
 			int swift_error_args = 0, swift_self_args = 0;
 			for (int i = 0; i < method->signature->param_count; ++i) {
 				MonoClass *param_klass = mono_class_from_mono_type_internal (method->signature->params [i]);
+				bool is_swift_self_generic = (strcmp(param_klass->name, "SwiftSelf`1") == 0) && (strcmp(param_klass->name_space, "System.Runtime.InteropServices.Swift") == 0);
```

Comment:
This is not the right way to check that `param_klass` is an instance of a known generic.  (you should not rely on the name/name_space of a generic instance being the same as the gtd; also you need to check that the image of the definition is CoreLib).

you want something like:

```c
GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL(swift_self_t);

GENERATE_TRY_GET_CLASS_WITH_CACHE(swift_self_t, "System.Runtime.InteropServices.Swift", "SwiftSelf`1");

gboolean
is_swift_self_inst (MonoClass *klass)
{
  MonoGenericClass *gklass = mono_class_try_get_generic_class (klass);
  if (!gklass)
    return FALSE;
  return gklass->container_class == mono_class_try_get_swift_self_t_class ();
}

Reply:
Thanks! Please let me know if you prefer a helper method (as you outlined). Since there are already many helper methods, I am wondering if this would improve readability.